File: swift/Sources/GloveboxValidator/Validator.swift
import CryptoKit
import Foundation

/// Wallet-side package validator. It mirrors the reference Python validator
/// check by check and reports the same error codes, so both implementations can
/// be tested differentially against the same vectors.
public struct Rejection: Error, Equatable {
    public let code: String
    public let detail: String
}

public struct ValidatedPackage {
    public let package: [String: JSONValue]
    public let kid: String
}

public struct WalletProfile {
    let schemas: Set<String>
    let handlers: [String: JSONValue]
    let publishers: [String: JSONValue]
    let versionFloors: [String: Int64]
    let limits: [String: Int64]
    let keys: [String: Curve25519.Signing.PublicKey]
    public let clock: Date

    public init(json data: Data) throws {
        guard let root = try StrictJSONParser.parse(data).object else { throw Rejection(code: "E_PROFILE", detail: "") }
        schemas = Set(root["schemas"]?.array?.compactMap { $0.string } ?? [])
        handlers = root["handlers"]?.object ?? [:]
        publishers = root["publishers"]?.object ?? [:]
        versionFloors = (root["versionFloors"]?.object ?? [:]).compactMapValues { $0.int }
        limits = (root["limits"]?.object ?? [:]).compactMapValues { $0.int }
        var keys: [String: Curve25519.Signing.PublicKey] = [:]
        for (kid, entry) in publishers {
            if let hex = entry.object?["publicKey"]?.string, let raw = Data(hexString: hex) {
                keys[kid] = try Curve25519.Signing.PublicKey(rawRepresentation: raw)
            }
        }
        self.keys = keys
        guard let clock = root["clock"]?.string.flatMap(Validator.parseTime) else {
            throw Rejection(code: "E_PROFILE", detail: "clock")
        }
        self.clock = clock
    }

    func locality(_ action: [String: JSONValue]) -> String? {
        guard let type = action["type"]?.string, let spec = handlers[type]?.object?["locality"] else { return nil }
        if let fixed = spec.string { return fixed }
        guard let rule = spec.object, let param = rule["byParam"]?.string,
              let value = action["params"]?.object?[param]?.string else { return nil }
        return rule["map"]?.object?[value]?.string
    }
}

public enum Validator {
    static let packageType = "glovebox-package+json"
    static let packageFields: Set<String> = ["schema", "id", "version", "floor", "notAfter", "initial", "states"]
    static let terminalKinds: Set<String> = ["success", "deny", "error"]

    /// Parses the fixed UTC form yyyy-MM-ddTHH:mm:ssZ without a locale-dependent formatter.
    static func parseTime(_ text: String) -> Date? {
        let b = Array(text.utf8)
        guard b.count == 20, b[4] == 0x2D, b[7] == 0x2D, b[10] == 0x54, b[13] == 0x3A, b[16] == 0x3A,
              b[19] == 0x5A else { return nil }
        func num(_ range: Range<Int>) -> Int? {
            var value = 0
            for i in range {
                guard (0x30...0x39).contains(b[i]) else { return nil }
                value = value * 10 + Int(b[i] - 0x30)
            }
            return value
        }
        guard let y = num(0..<4), let m = num(5..<7), let d = num(8..<10), let hh = num(11..<13),
              let mm = num(14..<16), let ss = num(17..<19), (1...12).contains(m), (1...31).contains(d),
              hh < 24, mm < 60, ss < 60 else { return nil }
        // Days from the civil calendar (H. Hinnant's algorithm).
        let yy = m <= 2 ? y - 1 : y
        let era = (yy >= 0 ? yy : yy - 399) / 400
        let yoe = yy - era * 400
        let doy = (153 * (m + (m > 2 ? -3 : 9)) + 2) / 5 + d - 1
        let doe = yoe * 365 + yoe / 4 - yoe / 100 + doy
        let days = era * 146_097 + doe - 719_468
        return Date(timeIntervalSince1970: TimeInterval(days * 86_400 + hh * 3600 + mm * 60 + ss))
    }

    static func reject(_ code: String, _ detail: String = "") -> Rejection {
        Rejection(code: code, detail: detail)
    }

    /// `floors` holds the highest signed floor this wallet has accepted per package identifier.
    public static func validate(_ raw: Data, profile: WalletProfile, now: Date? = nil,
                                floors: [String: Int64] = [:]) throws -> ValidatedPackage {
        let now = now ?? profile.clock
        guard raw.count <= Int(profile.limits["maxPackageBytes"] ?? 16384) else { throw reject("E_LIMITS") }

        // 1. Envelope.
        guard let text = String(data: raw, encoding: .ascii) else { throw reject("E_PARSE") }
        let parts = text.split(separator: ".", omittingEmptySubsequences: false).map(String.init)
        if parts.count == 2 || (parts.count == 3 && parts[2].isEmpty) { throw reject("E_SIG_MISSING") }
        guard parts.count == 3, let headerData = Data(base64URL: parts[0]),
              let payload = Data(base64URL: parts[1]), let signature = Data(base64URL: parts[2])
        else { throw reject("E_PARSE") }
        guard let header = try? StrictJSONParser.parse(headerData).object,
              header["typ"]?.string == packageType else { throw reject("E_PARSE") }

        // 2. Authenticity.
        guard header["alg"]?.string == "EdDSA" else { throw reject("E_SIG_INVALID") }
        guard let kid = header["kid"]?.string, let key = profile.keys[kid],
              let publisher = profile.publishers[kid]?.object else { throw reject("E_UNKNOWN_KEY") }
        let signingInput = Data((parts[0] + "." + parts[1]).utf8)
        guard key.isValidSignature(signature, for: signingInput) else { throw reject("E_SIG_INVALID") }

        // 3. Strict payload parsing.
        let parsed: JSONValue
        do { parsed = try StrictJSONParser.parse(payload) }
        catch StrictJSONError.duplicateKey(let key) { throw reject("E_DUPLICATE_KEY", key) }
        catch { throw reject("E_PARSE") }
        guard let package = parsed.object else { throw reject("E_PARSE") }

        // 4. Schema and namespace.
        guard let schema = package["schema"]?.string, profile.schemas.contains(schema) else { throw reject("E_SCHEMA") }
        guard Set(package.keys) == packageFields, let pid = package["id"]?.string,
              let version = package["version"]?.int, version >= 1,
              let floor = package["floor"]?.int, floor >= 1, floor <= version,
              let states = package["states"]?.object, let initial = package["initial"]?.string
        else { throw reject("E_SCHEMA") }
        let namespaces = publisher["namespaces"]?.array?.compactMap { $0.string } ?? []
        guard namespaces.contains(where: { pid.hasPrefix($0) }) else { throw reject("E_NAMESPACE") }

        // 5. Lifecycle.
        guard let notAfter = package["notAfter"]?.string.flatMap(parseTime) else { throw reject("E_SCHEMA") }
        if now > notAfter { throw reject("E_EXPIRED") }
        if version < max(profile.versionFloors[pid] ?? 1, floors[pid] ?? 1) { throw reject("E_ROLLBACK") }
        if states.count > Int(profile.limits["maxStates"] ?? 64) { throw reject("E_LIMITS") }

        // 6. Graph integrity.
        try checkGraph(states: states, initial: initial)

        // 7. Typing, capability, disclosure, endpoints, offline outcomes.
        let origins = Set(publisher["origins"]?.array?.compactMap { $0.string } ?? [])
        let granted = Set(publisher["handlers"]?.array?.compactMap { $0.string } ?? [])
        for sid in states.keys.sorted() {
            guard let state = states[sid]?.object, state["terminal"] == nil else { continue }
            guard let action = state["action"]?.object, let type = action["type"]?.string,
                  let spec = profile.handlers[type]?.object else { throw reject("E_UNKNOWN_ACTION", sid) }
            guard granted.contains(type) else { throw reject("E_PUBLISHER_SCOPE", sid) }
            guard Set(action.keys).isSubset(of: ["type", "params"]) else { throw reject("E_SCHEMA", sid) }
            try checkParams(sid: sid, params: action["params"] ?? .object([:]), spec: spec,
                            limits: profile.limits, origins: origins)
            let declared = Set(spec["outcomes"]?.array?.compactMap { $0.string } ?? [])
            let on = state["on"]?.object ?? [:]
            for outcome in on.keys where !declared.contains(outcome) { throw reject("E_UNDECLARED_OUTCOME", sid) }
            if profile.locality(action) == "network" && on["offline"] == nil { throw reject("E_OFFLINE_UNMAPPED", sid) }
        }
        return ValidatedPackage(package: package, kid: kid)
    }

    /// Validate against the stored floors, then ratchet the floor for this package identifier.
    @discardableResult
    public static func install(_ raw: Data, profile: WalletProfile, floors: inout [String: Int64],
                               now: Date? = nil) throws -> ValidatedPackage {
        let validated = try validate(raw, profile: profile, now: now, floors: floors)
        if let pid = validated.package["id"]?.string, let floor = validated.package["floor"]?.int {
            floors[pid] = max(floors[pid] ?? 1, floor)
        }
        return validated
    }

    static func checkGraph(states: [String: JSONValue], initial: String) throws {
        guard let first = states[initial]?.object, first["terminal"] == nil else { throw reject("E_BAD_INITIAL") }
        for sid in states.keys.sorted() {
            guard let state = states[sid]?.object else { throw reject("E_SCHEMA", sid) }
            if let terminal = state["terminal"] {
                guard state.count == 1, let kind = terminal.string, terminalKinds.contains(kind) else {
                    throw reject("E_TERMINAL", sid)
                }
                continue
            }
            guard Set(state.keys) == ["action", "on"], state["action"]?.object != nil else { throw reject("E_SCHEMA", sid) }
            guard let on = state["on"]?.object, !on.isEmpty else { throw reject("E_DANGLING", sid) }
            for target in on.values {
                guard let name = target.string, states[name] != nil else { throw reject("E_DANGLING", sid) }
            }
        }
        func successors(_ sid: String) -> [String] {
            (states[sid]?.object?["on"]?.object ?? [:]).values.compactMap { $0.string }
        }
        var reached: Set<String> = [initial]
        var frontier = [initial]
        while let sid = frontier.popLast() {
            for next in successors(sid) where !reached.contains(next) {
                reached.insert(next)
                frontier.append(next)
            }
        }
        if reached.count != states.count { throw reject("E_UNREACHABLE") }
        var colour: [String: Int] = [:]
        func visit(_ sid: String) throws {
            colour[sid] = 1
            for next in successors(sid) {
                if colour[next] == 1 { throw reject("E_CYCLE", sid) }
                if colour[next] == nil { try visit(next) }
            }
            colour[sid] = 2
        }
        try visit(initial)
    }

    static func checkParams(sid: String, params: JSONValue, spec: [String: JSONValue],
                            limits: [String: Int64], origins: Set<String>) throws {
        guard let params = params.object else { throw reject("E_PARAM", sid) }
        let rules = spec["params"]?.object ?? [:]
        if !Set(params.keys).isSubset(of: Set(rules.keys)) { throw reject("E_PARAM", sid) }
        let maxString = Int(limits["maxStringLength"] ?? 512)
        let maxList = Int(limits["maxListLength"] ?? 32)
        for name in rules.keys.sorted() {
            guard let rule = rules[name]?.object, let kind = rule["type"]?.string else { continue }
            var required = rule["required"] == .bool(true)
            if let condition = rule["requiredIf"]?.object {
                required = condition.allSatisfy { params[$0.key] == $0.value }
            }
            guard let value = params[name] else {
                if required { throw reject("E_PARAM", "\(sid): missing \(name)") }
                continue
            }
            switch kind {
            case "list":
                guard let items = value.array else { throw reject("E_PARAM", sid) }
                if rule["disclosure"] == .bool(true) && items.isEmpty { throw reject("E_EMPTY_DISCLOSURE", sid) }
                let strings = items.compactMap { $0.string }
                guard !items.isEmpty, items.count <= maxList, strings.count == items.count,
                      Set(strings).count == strings.count,
                      strings.allSatisfy({ !$0.isEmpty && $0.count <= maxString }) else { throw reject("E_PARAM", sid) }
            case "enum":
                let allowed = rule["values"]?.array ?? []
                guard allowed.contains(value) else { throw reject("E_PARAM", sid) }
            case "string":
                guard let s = value.string, !s.isEmpty, s.count <= maxString else { throw reject("E_PARAM", sid) }
            case "decimal":
                guard let s = value.string, isDecimal(s) else { throw reject("E_PARAM", sid) }
            case "endpoint":
                guard let s = value.string, let o = origin(s), origins.contains(o) else { throw reject("E_ENDPOINT", sid) }
            default:
                throw reject("E_PARAM", sid)
            }
        }
    }

    static func isDecimal(_ s: String) -> Bool {
        let parts = s.split(separator: ".", omittingEmptySubsequences: false)
        func digits(_ p: Substring, _ range: ClosedRange<Int>) -> Bool {
            range.contains(p.count) && p.allSatisfy { $0.isASCII && $0.isNumber }
        }
        if parts.count == 1 { return digits(parts[0], 1...6) }
        return parts.count == 2 && digits(parts[0], 1...6) && digits(parts[1], 1...2)
    }

    static func origin(_ url: String) -> String? {
        guard let c = URLComponents(string: url), c.scheme == "https", let host = c.host, !host.isEmpty,
              c.user == nil, c.password == nil else { return nil }
        return "https://" + host.lowercased() + (c.port.map { ":\($0)" } ?? "")
    }
}

extension Data {
    init?(base64URL text: String) {
        var s = text.replacingOccurrences(of: "-", with: "+").replacingOccurrences(of: "_", with: "/")
        if s.contains("=") { return nil }
        s += String(repeating: "=", count: (4 - s.count % 4) % 4)
        self.init(base64Encoded: s)
    }

    init?(hexString: String) {
        guard hexString.count % 2 == 0 else { return nil }
        var bytes: [UInt8] = []
        var index = hexString.startIndex
        while index < hexString.endIndex {
            let next = hexString.index(index, offsetBy: 2)
            guard let byte = UInt8(hexString[index..<next], radix: 16) else { return nil }
            bytes.append(byte)
            index = next
        }
        self.init(bytes)
    }
}

File: glovebox/validator.py
"""Package validation performed by the wallet before exposing any package UI.

A package travels as a compact JWS (RFC 7515) signed with EdDSA/Ed25519
(RFC 8037). The signature covers the exact payload bytes, so wallets written
in other languages need no shared canonicalization. The payload is parsed
strictly after the signature check: duplicate keys, floats, and non-finite
numbers are rejected.

Every rejection carries a stable error code so that the adversarial suite can
check that a mutation fails for the intended reason and not incidentally.
"""

import hashlib
import re
from urllib.parse import urlsplit

from . import ed25519
from .canonical import DuplicateKeyError, b64url_decode, strict_loads
from .profile import parse_time

TERMINAL_KINDS = ("success", "deny", "error")
PACKAGE_FIELDS = {"schema", "id", "version", "floor", "notAfter", "initial", "states"}
PACKAGE_TYPE = "glovebox-package+json"
DECIMAL_RE = re.compile(r"^[0-9]{1,6}(\.[0-9]{1,2})?$")


class Rejected(Exception):
    def __init__(self, code, detail):
        super().__init__(f"{code}: {detail}")
        self.code = code
        self.detail = detail


class ValidatedPackage:
    def __init__(self, package, kid, digest, size):
        self.package = package
        self.kid = kid
        self.digest = digest
        self.size = size


def origin(url):
    """Return the https origin of a URL, or None if the URL is not acceptable."""
    if not isinstance(url, str):
        return None
    try:
        parts = urlsplit(url)
        port = parts.port
    except ValueError:
        return None
    if parts.scheme != "https" or not parts.hostname or parts.username or parts.password:
        return None
    return f"https://{parts.hostname.lower()}" + (f":{port}" if port else "")


def endpoint_urls(package, profile):
    """All endpoint parameters of a package, as (state, parameter, url)."""
    found = []
    for sid, state in package["states"].items():
        if "action" not in state:
            continue
        spec = profile.handlers.get(state["action"].get("type"), {}).get("params", {})
        for name, value in state["action"].get("params", {}).items():
            if spec.get(name, {}).get("type") == "endpoint":
                found.append((sid, name, value))
    return found


def validate(raw, profile, now=None, floors=None):
    """Validate a serialized package; return a ValidatedPackage or raise Rejected.

    floors maps package identifiers to the highest floor this wallet has accepted
    (see install); without it, only the profile's configured floors apply.
    """
    now = now or profile.clock
    limits = profile.limits
    if len(raw) > limits["maxPackageBytes"]:
        raise Rejected("E_LIMITS", f"{len(raw)} bytes exceed the package size bound")

    # 1. Envelope: compact JWS with a protected header naming the publisher key.
    try:
        text = raw.decode("ascii")
    except UnicodeDecodeError:
        raise Rejected("E_PARSE", "a compact JWS is ASCII") from None
    parts = text.split(".")
    if len(parts) == 2 or (len(parts) == 3 and parts[2] == ""):
        raise Rejected("E_SIG_MISSING", "package carries no signature")
    if len(parts) != 3:
        raise Rejected("E_PARSE", "a compact JWS has three segments")
    try:
        header = strict_loads(b64url_decode(parts[0]))
        payload_bytes = b64url_decode(parts[1])
        signature = b64url_decode(parts[2])
    except (ValueError, UnicodeDecodeError) as exc:
        raise Rejected("E_PARSE", f"malformed JWS segment: {exc}") from None
    if not isinstance(header, dict) or header.get("typ") != PACKAGE_TYPE:
        raise Rejected("E_PARSE", "protected header must declare the package type")

    # 2. Authenticity: trusted key and a valid signature over the exact bytes.
    if header.get("alg") != "EdDSA":
        raise Rejected("E_SIG_INVALID", "only EdDSA (Ed25519) is accepted")
    kid = header.get("kid")
    if kid not in profile.publishers:
        raise Rejected("E_UNKNOWN_KEY", f"publisher key {kid!r} is not trusted")
    signing_input = (parts[0] + "." + parts[1]).encode("ascii")
    if not ed25519.verify(profile.public_key(kid), signing_input, signature):
        raise Rejected("E_SIG_INVALID", "signature does not verify")

    # 3. Strict payload parsing: parser differentials never reach later checks.
    try:
        package = strict_loads(payload_bytes)
    except DuplicateKeyError as exc:
        raise Rejected("E_DUPLICATE_KEY", f"duplicate JSON key {exc}") from None
    except (ValueError, UnicodeDecodeError) as exc:
        raise Rejected("E_PARSE", str(exc)) from None
    if not isinstance(package, dict):
        raise Rejected("E_PARSE", "payload must be a JSON object")

    # 4. Schema and namespace.
    if package.get("schema") not in profile.schemas:
        raise Rejected("E_SCHEMA", f"unsupported schema {package.get('schema')!r}")
    if set(package) != PACKAGE_FIELDS:
        raise Rejected("E_SCHEMA", "package fields do not match the schema")
    pid, version, pfloor = package["id"], package["version"], package["floor"]
    if not isinstance(pid, str) or any(isinstance(x, bool) or not isinstance(x, int) for x in (version, pfloor)):
        raise Rejected("E_SCHEMA", "malformed identifier, version, or floor")
    if not 1 <= pfloor <= version:
        raise Rejected("E_SCHEMA", "the floor must lie between 1 and the version")
    if version < 1 or not isinstance(package["states"], dict) or not isinstance(package["initial"], str):
        raise Rejected("E_SCHEMA", "malformed version, initial state, or state table")
    publisher = profile.publishers[kid]
    if not any(pid.startswith(ns) for ns in publisher["namespaces"]):
        raise Rejected("E_NAMESPACE", f"{kid} may not publish {pid}")

    # 5. Lifecycle: expiry and per-process version floor.
    try:
        not_after = parse_time(package["notAfter"])
    except ValueError:
        raise Rejected("E_SCHEMA", "malformed expiry") from None
    if now > not_after:
        raise Rejected("E_EXPIRED", f"package expired at {package['notAfter']}")
    floor = max(profile.version_floors.get(pid, 1), (floors or {}).get(pid, 1))
    if version < floor:
        raise Rejected("E_ROLLBACK", f"version {version} is below the floor {floor}")

    states = package["states"]
    if len(states) > limits["maxStates"]:
        raise Rejected("E_LIMITS", f"{len(states)} states exceed the state bound")

    # 6. Graph integrity: initial state, targets, reachability, acyclicity.
    _check_graph(package)

    # 7. Typing, capability confinement, disclosure, endpoints, offline outcomes.
    for sid, state in states.items():
        if "terminal" in state:
            continue
        action = state["action"]
        atype = action.get("type")
        if atype not in profile.handlers:
            raise Rejected("E_UNKNOWN_ACTION", f"{sid}: no native handler for {atype!r}")
        if atype not in publisher.get("handlers", []):
            raise Rejected("E_PUBLISHER_SCOPE", f"{sid}: {kid} is not granted {atype}")
        if set(action) - {"type", "params"}:
            raise Rejected("E_SCHEMA", f"{sid}: unexpected action members")
        _check_params(sid, atype, action.get("params", {}), profile, publisher)
        declared = profile.outcomes(atype)
        for outcome in state["on"]:
            if outcome not in declared:
                raise Rejected("E_UNDECLARED_OUTCOME", f"{sid}: {atype} never returns {outcome!r}")
        if profile.locality(action) == "network" and "offline" not in state["on"]:
            raise Rejected("E_OFFLINE_UNMAPPED", f"{sid}: network action lacks an offline transition")

    digest = hashlib.sha256(payload_bytes).hexdigest()
    return ValidatedPackage(package, kid, digest, len(raw))


def install(raw, profile, floors, now=None):
    """Validate against the wallet's stored floors, then ratchet the floor.

    The floor travels inside the signed payload, so only the publisher that may
    publish the package identifier can raise it. A wallet keeps the highest floor
    it has accepted; a forced update is a package whose floor equals its version.
    """
    validated = validate(raw, profile, now=now, floors=floors)
    pid = validated.package["id"]
    floors[pid] = max(floors.get(pid, 1), validated.package["floor"])
    return validated


def _check_graph(package):
    states = package["states"]
    initial = package["initial"]
    if initial not in states or not isinstance(states[initial], dict) or "terminal" in states[initial]:
        raise Rejected("E_BAD_INITIAL", f"initial state {initial!r} is missing or terminal")
    for sid, state in states.items():
        if not isinstance(state, dict):
            raise Rejected("E_SCHEMA", f"state {sid} is not an object")
        if "terminal" in state:
            if set(state) != {"terminal"} or state["terminal"] not in TERMINAL_KINDS:
                raise Rejected("E_TERMINAL", f"malformed terminal state {sid}")
            continue
        if set(state) != {"action", "on"} or not isinstance(state["action"], dict):
            raise Rejected("E_SCHEMA", f"malformed state {sid}")
        if not isinstance(state["on"], dict) or not state["on"]:
            raise Rejected("E_DANGLING", f"state {sid} has no outgoing transition")
        for outcome, target in state["on"].items():
            if target not in states:
                raise Rejected("E_DANGLING", f"{sid}.{outcome} targets missing state {target!r}")

    reached, frontier = {initial}, [initial]
    while frontier:
        sid = frontier.pop()
        for target in states[sid].get("on", {}).values():
            if target not in reached:
                reached.add(target)
                frontier.append(target)
    unreachable = sorted(set(states) - reached)
    if unreachable:
        raise Rejected("E_UNREACHABLE", f"unreachable states {unreachable}")

    back_edge = find_back_edge(initial, lambda sid: states[sid].get("on", {}).values())
    if back_edge:
        raise Rejected("E_CYCLE", f"cycle through {back_edge[0]} -> {back_edge[1]}")


def find_back_edges(initial, successors, first_only=False):
    """Iterative depth-first search; returns the back edges (cycles) reachable from initial."""
    colour = {initial: "grey"}
    stack = [(initial, iter(list(successors(initial))))]
    found = []
    while stack:
        node, pending = stack[-1]
        target = next(pending, None)
        if target is None:
            colour[node] = "black"
            stack.pop()
            continue
        state = colour.get(target)
        if state == "grey":
            found.append((node, target))
            if first_only:
                return found
        elif state is None:
            colour[target] = "grey"
            stack.append((target, iter(list(successors(target)))))
    return found


def find_back_edge(initial, successors):
    edges = find_back_edges(initial, successors, first_only=True)
    return edges[0] if edges else None


def _check_params(sid, atype, params, profile, publisher):
    spec = profile.handlers[atype]["params"]
    limits = profile.limits
    if not isinstance(params, dict):
        raise Rejected("E_PARAM", f"{sid}: parameters must be an object")
    unknown = sorted(set(params) - set(spec))
    if unknown:
        raise Rejected("E_PARAM", f"{sid}: unknown parameters {unknown}")
    for name, rule in spec.items():
        required = rule.get("required", False)
        condition = rule.get("requiredIf")
        if condition:
            required = all(params.get(key) == value for key, value in condition.items())
        if name not in params:
            if required:
                raise Rejected("E_PARAM", f"{sid}: missing parameter {name}")
            continue
        value = params[name]
        kind = rule["type"]
        if kind == "list":
            if not isinstance(value, list):
                raise Rejected("E_PARAM", f"{sid}: {name} must be a list")
            if rule.get("disclosure") and not value:
                raise Rejected("E_EMPTY_DISCLOSURE", f"{sid}: empty claim set")
            if not value or len(value) > limits["maxListLength"] or len(set(value)) != len(value):
                raise Rejected("E_PARAM", f"{sid}: {name} is empty, too long, or repeats items")
            if not all(isinstance(item, str) and item and len(item) <= limits["maxStringLength"]
                       for item in value):
                raise Rejected("E_PARAM", f"{sid}: {name} must contain bounded non-empty strings")
        elif kind == "enum":
            if value not in rule["values"]:
                raise Rejected("E_PARAM", f"{sid}: {name}={value!r} is not permitted")
        elif kind == "string":
            if not isinstance(value, str) or not value or len(value) > limits["maxStringLength"]:
                raise Rejected("E_PARAM", f"{sid}: {name} must be a bounded non-empty string")
        elif kind == "decimal":
            if not isinstance(value, str) or not DECIMAL_RE.match(value):
                raise Rejected("E_PARAM", f"{sid}: {name} must be a decimal amount")
        elif kind == "endpoint":
            if origin(value) not in publisher["origins"]:
                raise Rejected("E_ENDPOINT", f"{sid}: {name}={value!r} is not a registered origin")
        else:
            raise Rejected("E_PARAM", f"{sid}: unsupported parameter type {kind}")
